types/validation.go

```diff
@@ -298,22 +356,44 @@ func verifyCommitBatch(
 		return ErrNotEnoughVotingPowerSigned{Got: got, Needed: needed}
 	}
 
+	// if every signature was in the cache, the batch verifier is empty and we shouldn't call verify
+	if len(batchSigIdxs) == 0 {
+		return nil
+	}
+
 	// attempt to verify the batch.
 	ok, validSigs := bv.Verify()
 	if ok {
 		// success
+		if verifiedSignatureCache != nil {
+			for i := range validSigs {
+				idx := batchSigIdxs[i]
+				sig := commit.Signatures[idx]
+				verifiedSignatureCache.Add(string(sig.Signature), SignatureCacheValue{
+					ValidatorAddress: sig.ValidatorAddress,
+					VoteSignBytes:    commit.VoteSignBytes(chainID, int32(idx)),
```

Comment:
This call requires proto encoding, namely is kind of expensive. Don't we have these bytes already computed?